Utils/ObjectPool.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace DUCK.Pooling
+{
+	public static class ObjectPool
+	{
+		private interface Pool
+		{
+			Type ObjectType { get; }
+
+			void Add(PoolableObject obj);
+		}
+
+		private class Pool<T> : Pool 
+			where T : PoolableObject
+		{
+			private Queue<T> pooledObjects = new Queue<T>();
+
+			public Type ObjectType
+			{
+				get
+				{
+					return typeof(T);
+				}
+			}
+
+			public T Get()
+			{
+				T obj = null;
+				while (pooledObjects.Count > 0)
+				{
+					obj = pooledObjects.Dequeue();
+				}
+
+				if (obj)
+				{
+					obj.RemoveFromPool();
+				}
+
+				return obj;
+			}
+
+			public void Add(PoolableObject obj)
+			{
+				if (obj != null && !pooledObjects.Contains((T)obj))
+				{
+					obj.AddToPool();
+					pooledObjects.Enqueue((T)obj);
+				}
+			}
+		}
+
+		private static Dictionary<Type, Pool> poolLookup = new Dictionary<Type, Pool>();
+
+		public static T Instantiate<T>(T original = null) where T : PoolableObject
+		{
+			T obj = null;
+
+			if (PoolExists(typeof(T)))
+			{
+				obj = GetFromPool<T>();
+
+				if (obj != null)
+				{
+					if (original != null)
+					{
+						obj.CopyFrom(original);
+					}
+
+					return obj;
+				}
+			}
+			else
+			{
+				CreatePool<T>();
+			}
+
+			return obj ?? GameObject.Instantiate<T>(original);
```

Comment:
Either line 72 needs removing or this line needs to be `return GameObject.Instantiate<T>(original);`. The reference to `obj` will always be null if the program reaches here.